### ISSUE TITLE
Backtesting results don't match hyperopted protections

### ISSUE THREAD

**Issue by infiniteneurotech** [freqtrade/freqtrade#7638]
## Describe your environment

  * Operating system: Windows 10
  * Python Version: 3.10.8 (`python -V`)
  * CCXT version: 1.93.98 (`pip freeze | grep ccxt`)
  * Freqtrade Version: 2022.8 (`freqtrade -V` or `docker-compose run --rm freqtrade -V` for Freqtrade running in docker)
  
## Your question

Hello,

I ran a hyperopt on a strategy and I can't get the backtest results to match the hyperopt results. I only hyperopted a low profit protection parameter (--spaces protections), and after running a backtest (on the same file, with --enable-protections) the results were very different. I checked the .json file and the protection settings matched the hyperopt results so I can't figure out what's going wrong. Here are the hyperopt commands from the file:

``` python
stop_duration = IntParameter(1, 500, default=48, space='protection', optimize=True)
required_profit = DecimalParameter(0.005, 10, default=0.01, space='protection', optimize=True)
look_back = IntParameter(1, 500, default=200, space='protection', optimize=True)
use_stop_protection = BooleanParameter(default=True, space="protection", optimize=True)
```

``` python 
    @property
    def protections(self):
        prot = []

        
        if self.use_stop_protection.value:
            prot.append({
                "method": "LowProfitPairs",
                "lookback_period_candles": self.look_back.value,
                "trade_limit": 2,
                "stop_duration_candles": self.stop_duration.value,
                "required_profit": self.required_profit.value,
                "only_per_pair": False,
            })

        return prot
```

These were the results:

"protection": {
      "look_back": 12,
      "required_profit": 9.387,
      "stop_duration": 157,
      "use_stop_protection": true
    }

Any advice would be appreciated, thanks

**Comment by xmatthias** [comment]
from the small excerpt, i don't see anything wrong.

we can however only really assist if we have a working example (a minimal strategy that reproduces the problem you're having) -  best including the precise commands that were used
e.g. `freqtrade hyperopt ... `, `freqtrade backtesting ...` (where `...` is the actually interesting part).

Often, only the precise commands used can reveal something interesting - so maybe let's start with these.

**Comment by infiniteneurotech** [comment]
Ok thanks for your reply. My commands were:

freqtrade hyperopt -s Ichi_hyperopt_1 --timerange 20200701-20221024 --spaces protections -e 200 --hyperopt-loss SharpeHyperOptLoss

freqtrade backtesting -s Ichi_hyperopt_1 --timerange 20200701-20221024 --enable-protections

The strategy runs on a 1h timeframe, defined in the config file. No ROI or stoploss set, and as you can see above, only hyperopting the protections space.

My full strategy is quite big (and is working well, aside from this hyperopt anomaly) so I'll try and replicate the problem on one of my simpler ones and I'll attach it shortly.

Thanks

**Comment by infiniteneurotech** [comment]
While trying to replicate this problem in another strategy I was just reminded of another problem I've been having. Almost every time I run hyperopt it causes my Windows Linux service to crash, and I get an error saying Freqtrade can't connect to Binance. I have to shutdown WSL and restart Docker, which fixes it temporarily. Because of that crash I'm now not 100% certain if these results are a mismatch as they're in the same range.

Anyway, not trying to hijack my own question. Here's the strategy:
```
# pragma pylint: disable=missing-docstring, invalid-name, pointless-string-statement
# flake8: noqa: F401
# isort: skip_file
# --- Do not remove these libs ---
import numpy as np  # noqa
import pandas as pd  # noqa
from pandas import DataFrame
import pandas_ta as ta
from datetime import datetime
from freqtrade.persistence import Trade
from technical.util import resample_to_interval, resampled_merge
from functools import reduce

from freqtrade.strategy import (BooleanParameter, CategoricalParameter, DecimalParameter,
                                IStrategy, IntParameter, stoploss_from_absolute, merge_informative_pair)

# --------------------------------
# Add your lib to import here
import talib.abstract as ta
import freqtrade.vendor.qtpylib.indicators as qtpylib
import pandas_ta as pta



# This class is a sample. Feel free to customize it.
class Ichi_2(IStrategy):
   
    # Can this strategy go short?
    can_short: bool = False

    custom_info = {
        'risk_reward_ratio': 5,
        'set_to_break_even_at_profit': 5,
    }
    stoploss = -0.999 #-0.137
    minimal_roi = {
        "0": 999
    }
    

    # Trailing stoploss
    trailing_stop = False #set to false 
    trailing_only_offset_is_reached = False
    trailing_stop_positive = 0.266
    trailing_stop_positive_offset = 0.36  # Disabled / not configured

    # Optimal timeframe for the strategy.
    timeframe = '1h'

    # Run "populate_indicators()" only for new candle.
    process_only_new_candles = False

    # These values can be overridden in the config.
    use_exit_signal = True
    exit_profit_only = False
    ignore_roi_if_entry_signal = False
    use_custom_stoploss = False


     # Hyperopt spaces
    
    stop_duration = IntParameter(1, 500, default=48, space='protection', optimize=True)
    required_profit = DecimalParameter(0.005, 10, default=0.01, space='protection', optimize=True)
    look_back = IntParameter(1, 500, default=200, space='protection', optimize=True)
    use_stop_protection = BooleanParameter(default=True, space="protection", optimize=True)


    startup_candle_count: int = 53

    

    # Optional order type mapping.
    order_types = {
        'entry': 'market',
        'exit': 'market',
        'stoploss': 'market',
        'stoploss_on_exchange': False
    }

    # Optional order time in force.
    order_time_in_force = {
        'entry': 'gtc',
        'exit': 'gtc'
    }

    

    @property
    def protections(self):
        prot = []

        
        if self.use_stop_protection.value:
            prot.append({
                "method": "LowProfitPairs",
                "lookback_period_candles": self.look_back.value,
                "trade_limit": 2,
                "stop_duration_candles": self.stop_duration.value,
                "required_profit": self.required_profit.value,
                "only_per_pair": False,
            })

        return prot

    def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        if not self.dp:
            # Don't do anything if DataProvider is not available.
            return dataframe

        
       
        dataframe['atr'] = ta.ATR(dataframe, timeperiod=10)
        
       

        #Trailing ATR exit
        dataframe['trail_low'] = dataframe['high'] - (dataframe['atr'] * 2)
        dataframe['highest'] = dataframe['trail_low'].rolling(10).max()
        dataframe['exit_signal'] = pd.concat([dataframe['trail_low'], dataframe['highest'].shift(1)], axis=1).max(1) #sell signal can only ever increase

        # Ichimoku
        # Regular settings:
        # Conversion line rolling window = 9
        # Baseline rolling window = 26
        # Leading span b rolling window = 52
        # Span A and B shift 26 periods
        # Chikou shift -26 periods

        # Crypto settings:
        # Conversion line rolling window = 20
        # Baseline rolling window = 60
        # Leading span b rolling window = 120
        # Span A and B shift 60 periods
        # Chikou shift -60 periods

        dataframe['conversion_line'] = (dataframe['high'].rolling(window=9).max()
            + dataframe['low'].rolling(window=9).min()) / 2

        dataframe['baseline'] = (dataframe['high'].rolling(window=26).max()
            + dataframe['low'].rolling(window=26).min()) / 2

        dataframe['leading_span_a'] = (dataframe['conversion_line'] + dataframe['baseline']) / 2

        dataframe['leading_span_b'] = (dataframe['high'].rolling(window=52).max()
                            + dataframe['low'].rolling(window=52).min()) / 2

        dataframe['span_a'] = dataframe['leading_span_a'].shift(26)

        dataframe['span_b'] = dataframe['leading_span_b'].shift(26)

        #dataframe['chikou_span'] = dataframe['close'].shift(-26)

        dataframe['cloud_green'] = (dataframe['span_a'] > dataframe['span_b'])
        dataframe['cloud_red'] = (dataframe['span_b'] > dataframe['span_a'])

        
        return dataframe



        
    def populate_entry_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        

             

        dataframe.loc[
            (
                qtpylib.crossed_above(dataframe['close'], dataframe['baseline']) &
                (dataframe['cloud_green'] == True) &
                
                                
                (dataframe['volume'] > 0)
                
            ),

            ['enter_long', 'enter_tag']] = (1, "Baseline_cross_cloud")

        
        return dataframe




    def populate_exit_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        
        
        dataframe.loc[
            (               
                
                qtpylib.crossed_below(dataframe['close'], dataframe['exit_signal']) &
                
                (dataframe['volume'] > 0) 
                
                
                
            ),

            ['exit_long', 'exit_tag']] = (1, 'Trailing ATR')

        
        
               
        return dataframe
```


It's late here and I have work in the morning, but I'll check back tomorrow. Thanks for taking the time to check.

**Comment by xmatthias** [comment]
For now - i can't really recreate what you're seeing.

on a (random) pairlist/config i had lying around and happened to have data for:

backtesting:
```
================== SUMMARY METRICS ==================
| Metric                      | Value               |
|-----------------------------+---------------------|
| Backtesting from            | 2020-07-29 05:00:00 |
| Backtesting to              | 2021-12-25 20:00:00 |
| Max open trades             | 2                   |
|                             |                     |
| Total/Daily Avg Trades      | 255 / 0.5           |
| Starting balance            | 1000 USD            |
| Final balance               | 1096.985 USD        |
| Absolute profit             | 96.985 USD          |
| Total profit %              | 9.70%               |
| CAGR %                      | 6.79%               |
| Profit factor               | 1.42                |
| Trades per day              | 0.5                 |
| Avg. daily profit %         | 0.02%               |
| Avg. stake amount           | 49.725 USD          |
| Total trade volume          | 12679.754 USD       |

```

hyperopt - you can get to this by using `hyperopt-show -n 5` (epoch 5 was my best epoch) - and it'll also write the parameters again "just to be sure":
```

================== SUMMARY METRICS ==================                                                                               
| Metric                      | Value               |                                                                               
|-----------------------------+---------------------|                                                                               
| Backtesting from            | 2020-07-29 05:00:00 |                                                                               
| Backtesting to              | 2021-12-25 20:00:00 |                                                                               
| Max open trades             | 2                   |                                                                               
|                             |                     |                                                                               
| Total/Daily Avg Trades      | 255 / 0.5           |                                                                               
| Starting balance            | 1000 USD            |                                                                               
| Final balance               | 1096.985 USD        |                                                                               
| Absolute profit             | 96.985 USD          |                                                                               
| Total profit %              | 9.70%               |                                                                               
| CAGR %                      | 6.79%               |                                                                               
| Profit factor               | 1.42                |                                                                               
| Trades per day              | 0.5                 |                                                                               
| Avg. daily profit %         | 0.02%               |                                                                               
| Avg. stake amount           | 49.725 USD          |                                                                               
| Total trade volume          | 12679.754 USD       |   
```

**Comment by infiniteneurotech** [comment]
Thanks for looking into this for me. I'm trying again to replicate the problem this morning, and now even with the original problematic strategy it's working as it should. I didn't change anything at all, and left my computer on overnight. I just followed the same procedure that was causing the problem last night and now it's working perfectly. One thing I noticed is that in the output I'm now seeing 2022-10-24 22:45:20,784 - freqtrade.resolvers.iresolver - INFO - Using resolved protection LowProfitPairs from '/freqtrade/freqtrade/plugins/protections/low_profit_pairs.py'... which wasn't there last night. I had wondered whether the backtesting module just wasn't using the protections, but when I ran the backtest without --enable-protections specified the results were different from what I was getting with --enable-protections, but also different from the hyperopt results.

This is a bit of a mystery to me still, but I'm glad it's working. If I can provide any further information to help with debugging please let me know. Thanks again for your help.

**Comment by xmatthias** [comment]
> Almost every time I run hyperopt it causes my Windows Linux service to crash

i suspect the key for this is in the above.
if hyperopt isn't finishing normally, it'll not write the parameters file (only the epochs themselves). 
you can force writing of the parameter file by running `hyperopt-show -n <epoch>`.

**Comment by infiniteneurotech** [comment]
Interesting, thanks for that. I tried running the hyperopt-show command but got an error saying the command wasn't recognised. I had a very brief search of the documentation to find out why, but since the backtests were working I didn't pursue it. It appears to me that the real issue is WSL crashing anyway so I'll look into how to fix that and that should solve the parameter writing issue. Thanks again for your help!
